### ISSUE TITLE
systemd-shipped units goes into failed state due to rate-limiting checks being performed earlier

### ISSUE THREAD

**Issue by YHNdnzj** [systemd/systemd#21025]
**systemd version the issue has been seen with**
 > 249.5

**Used distribution**
 > Arch Linux

`systemd-{binfmt,hwdb-update,sysusers,firstboot,repart}.service` and `systemd-ask-password-console.path` have been failing for me in 249.5, which would enter `inactive(dead)` state because of condition fail in older versions. Looking into the commit history, it seems that rate-limiting is now checked before conditions (#20531). And for some reason, the mentioned services got started multiple times during boot, resulting in hitting the rate limit and thus going into failed state.

Are there any ways to investigate why these units are started more than once (I've checked no other units explicitly depend on them), or should it be reverted to performing condition checks first?

([The manual of system.unit](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#StartLimitIntervalSec=interval) also suggests that "Rate-limiting is enforced after any unit condition checks are executed, and hence unit activations with failing conditions do not count towards the rate limit." It may need updating.)

**Comment by sluops** [comment]
It currently looks like this after boot:
```
[root@athena ~]► systemctl list-units --state=failed
  UNIT                        LOAD   ACTIVE SUB    DESCRIPTION                 >
● systemd-binfmt.service      loaded failed failed Set Up Additional Binary For>
● systemd-firstboot.service   loaded failed failed First Boot Wizard
● systemd-fsck-root.service   loaded failed failed File System Check on Root De>
● systemd-hwdb-update.service loaded failed failed Rebuild Hardware Database
● systemd-repart.service      loaded failed failed Repartition Root Disk
● systemd-sysusers.service    loaded failed failed Create System Users
```
Exemplary on `systemd-binfmt.service` shows that the start request ran against the rate limit:
```
[root@athena ~]► systemctl status systemd-binfmt.service 
× systemd-binfmt.service - Set Up Additional Binary Formats
     Loaded: loaded (/usr/lib/systemd/system/systemd-binfmt.service; static)
     Active: failed (Result: start-limit-hit)
  Condition: start condition failed at Mon 2021-10-18 00:12:40 CEST; 5min ago
             ├─ ConditionDirectoryNotEmpty=|/lib/binfmt.d was not met
             ├─ ConditionDirectoryNotEmpty=|/usr/lib/binfmt.d was not met
             ├─ ConditionDirectoryNotEmpty=|/usr/local/lib/binfmt.d was not met
             ├─ ConditionDirectoryNotEmpty=|/etc/binfmt.d was not met
             └─ ConditionDirectoryNotEmpty=|/run/binfmt.d was not met
       Docs: man:systemd-binfmt.service(8)
             man:binfmt.d(5)
             https://www.kernel.org/doc/html/latest/admin-guide/binfmt-misc.html
             https://www.freedesktop.org/wiki/Software/systemd/APIFileSystems

Oct 18 00:12:39 athena systemd[1]: Condition check resulted in Set Up Additional Binary Formats being skipped.
Oct 18 00:12:39 athena systemd[1]: Condition check resulted in Set Up Additional Binary Formats being skipped.
Oct 18 00:12:40 athena systemd[1]: Condition check resulted in Set Up Additional Binary Formats being skipped.
Oct 18 00:12:40 athena systemd[1]: Condition check resulted in Set Up Additional Binary Formats being skipped.
Oct 18 00:12:41 athena systemd[1]: systemd-binfmt.service: Start request repeated too quickly.
Oct 18 00:12:41 athena systemd[1]: systemd-binfmt.service: Failed with result 'start-limit-hit'.
Oct 18 00:12:41 athena systemd[1]: Failed to start Set Up Additional Binary Formats.
```

```
[root@athena ~]► systemctl cat systemd-binfmt.service 
# /usr/lib/systemd/system/systemd-binfmt.service
#  SPDX-License-Identifier: LGPL-2.1-or-later
#
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.

[Unit]
Description=Set Up Additional Binary Formats
Documentation=man:systemd-binfmt.service(8) man:binfmt.d(5)
Documentation=https://www.kernel.org/doc/html/latest/admin-guide/binfmt-misc.html
Documentation=https://www.freedesktop.org/wiki/Software/systemd/APIFileSystems
DefaultDependencies=no
Conflicts=shutdown.target
After=proc-sys-fs-binfmt_misc.automount
After=proc-sys-fs-binfmt_misc.mount
Before=sysinit.target shutdown.target
ConditionPathIsReadWrite=/proc/sys/
ConditionDirectoryNotEmpty=|/lib/binfmt.d
ConditionDirectoryNotEmpty=|/usr/lib/binfmt.d
ConditionDirectoryNotEmpty=|/usr/local/lib/binfmt.d
ConditionDirectoryNotEmpty=|/etc/binfmt.d
ConditionDirectoryNotEmpty=|/run/binfmt.d

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStart=/usr/lib/systemd/systemd-binfmt
ExecStop=/usr/lib/systemd/systemd-binfmt --unregister
TimeoutSec=90s
```

**Comment by loqs** [comment]
From the downstream bug report on for Arch Linux [1] reverting https://github.com/systemd/systemd-stable/commit/ed8fbbf1745c6a2dc0b8cd560ac8a3353f72e979 restores the previous behavior.

@DaanDeMeyer is the change in behavior intended?

[1]:  https://bugs.archlinux.org/task/72425

**Comment by yuwata** [comment]
Note, v248-stable repository already reverted the offending patch.
https://github.com/systemd/systemd-stable/commit/f0bce95d03cfdc3b190aef80029eccc36261616e

I've opened a PR which also revert the patch for v249.
https://github.com/systemd/systemd-stable/pull/133

**Comment by maxyu1115** [comment]
@DaanDeMeyer Sorry I know this issue is outdated, but we just upgraded to v250.5, and noticed a variation of this. 

What was the rationale to only delay the check https://github.com/systemd/systemd/commit/ce2146f5256659c7fb53a7d5b9dc551252e27e7e, versus a full revert of https://github.com/systemd/systemd/pull/20531? Asking because after these two changes, the check is still moved earlier.

We essentially noticed a bug where we enter the code path https://github.com/DaanDeMeyer/systemd/blob/9727f2427ff6b2e1f4ab927cc57ad8e888f04e95/src/core/service.c#L2435, where EAGAIN let's systemd restart our service later, but this counts twice against the start limit. (So for example with StartLimitBurst=3, we used to get 3 restart attempts, but now only 2.)

I don't really understand why we go down that code path, but we noticed the behavior difference, and wanted to discuss what we should be doing. We are currently only triggering this with a very specific test that causes the service to core dump, but are worried about how else this can happen.

Thanks!
